// nimma/cjs/codegen/templates/scope.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var builders = require('../ast/builders.cjs');

const SCOPE_IDENTIFIER = builders.identifier('scope');

const PATH = builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('path'));
const DEPTH = builders.memberExpression(PATH, builders.identifier('length'));

var scope = {
  _: SCOPE_IDENTIFIER,

  allocState: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('allocState')),
  callbacks: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('callbacks')),
  depth: DEPTH,
  destroy: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('destroy')),
  emit: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('emit')),
  fork: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('fork')),
  path: PATH,
  property: builders.memberExpression(
    PATH,
    builders.binaryExpression('-', DEPTH, builders.numericLiteral(1)),
    true,
  ),
  sandbox: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('sandbox')),
  shorthands: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('shorthands')),
  traverse: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('traverse')),
  value: builders.memberExpression(SCOPE_IDENTIFIER, builders.identifier('value')),
};

exports.default = scope;
